### ISSUE TITLE
Delay in `probability_that_sequence_is_increasing`

### ISSUE THREAD

**Issue by hbredin** [davisking/dlib#1257]
Thanks for the great library!

I am trying to implement "backtracking" as suggested in your [latest blog post](http://blog.dlib.net/2018/02/automatic-learning-rate-scheduling-that.html).

To do that, I am using a code similar to the following:

```python
>>> losses = []
>>> for batch in training_batches:
...     batch_loss = train(batch)
...     losses.append(batch_loss)
...     if dlib.probability_that_sequence_is_increasing(losses) > 0.99:
...         backtrack()
```

However, I noticed a significant delay between when the loss is actually starting to increase, and when the probability actually increaases to 1. as well. In the above example, around 13000th batch, more than 1000 batches pass before probability_that_sequence_is_increasing notices it.

![probability_that_sequence_is_increasing](https://user-images.githubusercontent.com/1080837/38816777-2675678a-4197-11e8-93ec-a057664c4e18.jpeg)

Is this the intended behavior? Am I doing something wrong?

**Comment by davisking** [comment]
It's not going to instantaneously go to 1. That wouldn't make sense.  You
have a huge history window, most of which is clearly decreasing and you ask
"is the loss decreasing in the window?" and the answer is yes until the
window as a whole isn't decreasing. Think about the model in the blog
post.  It fits a line and asks about the slope of the line.  The line is
fit to the entire window. You need to use a window of an appropriate length.

**Comment by hbredin** [comment]
OK. I did not understand that it would use the entire history.
That makes sense.  Thanks for the clarification.

I'll switch to a history containing approximately the number of batches in one epoch.
The hypothesis would then be "did the loss increase during the last epoch?".

**Comment by davisking** [comment]
Right.
